src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java

```diff
@@ -92,63 +111,116 @@ protected void init(MAPF_Instance instance, RunParameters parameters) {
         initPriority(instance);
 
         // distance between every vertex in the graph to each agent's goal
-        this.heuristic = new DistanceTableAStarHeuristic(instance.agents, instance.map);
+        if (parameters.aStarGAndH instanceof DistanceTableAStarHeuristic) {
+            this.heuristic = (DistanceTableAStarHeuristic) parameters.aStarGAndH;
+        }
+        else {
+            this.heuristic = new DistanceTableAStarHeuristic(instance.agents, instance.map);
+        }
     }
 
 
     @Override
     protected Solution runAlgorithm(MAPF_Instance instance, RunParameters parameters) {
-
+        boolean loopDetected = false;
         // each iteration of the while represents timestamp
         while (!(finished())) {
 
+            // algorithm starts agent's plans only from problemTimeStamp
+//            if (this.problemStartTime > this.timeStamp) {
+//                this.timeStamp++;
+//                continue;
+//            }
+
+            // loop?
+            if (this.timeStamp > 0) {
+                for (int i = 2; i < this.timeStamp; i++) {
+                    int agentsInSameLocation = 0;
+                    for (Agent agent : this.agentPlans.keySet()) {
+                        SingleAgentPlan currentPlan = this.agentPlans.get(agent);
+                        // check whether locations repeat themselves
+                        if (currentPlan.moveAt(i).currLocation.equals(currentPlan.getLastMove().currLocation)) {
+                            // if locations repeats themselves, check if statuses also repeats themselves
+                            if (this.reachingGoalsConfigurations.get(i).get(agent).equals(this.reachingGoalsConfigurations.get(this.timeStamp).get(agent))) {
+                                // agent could be in a loop
+                                agentsInSameLocation++;
+                            }
+                        }
+                    }
+//                    System.out.println("Agents in same location: " + agentsInSameLocation);
+                    if (agentsInSameLocation == this.agentPlans.keySet().size()) {
+                        System.out.println("Loop detected in timestamp: " + i);
+                        System.out.println("current timestamp: " + this.timeStamp);
```

Comment:
```suggestion
                        if (DEBUG >= 2){
                            System.out.println("Loop detected in timestamp: " + i);
                            System.out.println("current timestamp: " + this.timeStamp);
                        }
```
Don't print by default. Add `private static final int DEBUG = 0;` at the beginning of the class. But you can add it to the instance report if you want.